File: 764 Largest Plus Sign/764 Largest Plus Sign.go
func orderOfLargestPlusSign(n int, mines [][]int) int {
	grid := make([][]int, 0, n)
	for i := 0; i < n; i++ {
		row := make([]int, n, n)
		for j := 0; j < n; j++ {
			row[j] = n
		}
		grid = append(grid, row)
	}
	for _, ij := range mines {
		grid[ij[0]][ij[1]] = 0
	}

	for i := 0; i < n; i++ { // check four direction in one loop
		for j, left := 0, 0; j < n; j++ { // left
			if grid[i][j] == 0 {
				left = 0
			} else {
				left++
			}
			if left < grid[i][j] {
				grid[i][j] = left
			}
		}
		for j, right := n-1, 0; j >= 0; j-- { // right
			if grid[i][j] == 0 {
				right = 0
			} else {
				right++
			}
			if right < grid[i][j] {
				grid[i][j] = right
			}
		}
		for j, up := 0, 0; j < n; j++ { // up
			if grid[j][i] == 0 {
				up = 0
			} else {
				up++
			}
			if up < grid[j][i] {
				grid[j][i] = up
			}
		}
		for j, down := n-1, 0; j >= 0; j-- { // down
			if grid[j][i] == 0 {
				down = 0
			} else {
				down++
			}
			if down < grid[j][i] {
				grid[j][i] = down
			}
		}
	}

	r := 0
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] > r {
				r = grid[i][j]
			}
		}
	}

	return r
}

File: 764 Largest Plus Sign/764 Largest Plus Sign.py
"""
You are given an integer n. You have an n x n binary grid grid with all values initially 1's except for some indices given in the array mines. The ith element of the array mines is defined as mines[i] = [xi, yi] where grid[xi][yi] == 0.

Return the order of the largest axis-aligned plus sign of 1's contained in grid. If there is none, return 0.

An axis-aligned plus sign of 1's of order k has some center grid[r][c] == 1 along with four arms of length k - 1 going up, down, left, and right, and made of 1's. Note that there could be 0's or 1's beyond the arms of the plus sign, only the relevant area of the plus sign is checked for 1's.

Example 1:

Input: n = 5, mines = [[4,2]]
Output: 2
Explanation: In the above grid, the largest plus sign can only be of order 2. One of them is shown.

Example 2:

Input: n = 1, mines = [[0,0]]
Output: 0
Explanation: There is no plus sign, so return 0.
"""
class Solution:
    def orderOfLargestPlusSign(self, n: int, mines: [[int]]) -> int:
        cols = [[-1,n] for _ in range(n)]
        rows = [[-1,n] for _ in range(n)]
        for r,c in mines:
            cols[c].append(r)
            rows[r].append(c)
        for i in range(n):
            cols[i].sort()
            rows[i].sort()
        res = 0
        for r in range(n):
            for i in range(len(rows[r])-1):
                left = rows[r][i]+1
                right = rows[r][i+1]-1
                for c in range(left+res, right-res+1):
                    idx = bisect.bisect_right(cols[c],r)
                    up = cols[c][idx-1]+1
                    down = cols[c][idx]-1
                    # print(r,c,left,right,up,down)
                    origRes = res
                    res = max(res, min(c-left+1,right-c+1,r-up+1, down-r+1))
        return res
